runtime/onert/core/src/odc/CodegenManager.cc

```diff
@@ -0,0 +1,54 @@
+/*
+ * Copyright (c) 2024 Samsung Electronics Co., Ltd. All Rights Reserved
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "CodegenLoader.h"
+#include "odc/CodegenManager.h"
+#include "util/Utils.h"
+
+#include <fstream>
+#include <mutex>
+
+namespace onert
+{
+namespace odc
+{
+
+bool CodegenManager::codegen(const char *target, CodegenPreference pref)
+{
+  // codegen function is thread-unsafe
```

Comment:
Q. What does it mean?
1. This is thread-unsafe because of using the following `mutex`?
2. or because this is thread-unsafe, `mutex` should be used?

Reply:
Our internal code generator implementation was (at least as of my writing this code in internal repo) not thread-safe.

Reply:
(Optional)
How about moving this `std::mutex` to member variable without `static`? It would be OK since this function is a member function of `CodegenManager`. But I'm not sure if this member function will be thread-safe.

Reply:
@ragmani 
IMO, this mutex should be static variable. The static variable ensures that it is not executed by other functions during initialization. 

> If control enters the declaration concurrently while the variable is being initialized, the concurrent execution shall wait for completion of the initialization.

from https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2012/n3337.pdf §6.7

Reply:
> IMO, this mutex should be static variable. The static variable ensures that it is not executed by other functions during initialization.

This function is a member function, not non-member function. So, this member function can be called by an object of `CodegenManager`. Nevertheless, should we use `mutex` as a static variable?

Reply:
@glistening 

Could you let us know what the code is not thread-safe?

Reply:
@ragmani 

> This function is a member function, not non-member function. So, this member function can only be called by an object of `CodegenManager`. Nevertheless, should we use `mutex` as a static variable?

Regardless this function is a member function, this `codegen` function can be called at the same time in multiple threads. I think it considers this case.

Reply:
@jyoungyun 

> Regardless of whether this function is a member function, this `codegen` function can be called from multiple threads at the same time. I think it considers this case. 

Regardless of whether `std::mutex` variable is `static`, `std::mutex` guarantees thread-safe of the code below `std::lock_guard<std::mutex>` in the block. The `std::mutex` variable can be shared in all objects of `CodegenManager`. We only need to ensure that each object of `CodegenManger` is thread-safe if `CodegenManager::codegen()` is not thread-safe without using `std::mutex`.

Reply:
@jyoungyun 
As we discussed offline, I understand when `static std::mutex` is needed in the current code.

```mermaid
stateDiagram-v2
state Thread1 {
   CodegenManager1
}
CodegenManager1 --> CodegenLoader(Singleton) : load() & unload() continuously

state Thread2 {
   CodegenManager2
}

CodegenManager2 --> CodegenLoader(Singleton) : load() & unload() continuously
```

When two or more codegen managers try to concurrently call sequential functions(`CodegenLoader::loadLibrary()` and `CodegenLoader::unloadLibrary()`) " in each thread, the `std::mutex` variable should be `static` to avoid to ensure the atomicity of the members of `CodegenLoader` since `CodegenLoader` is singleton and its object is shared.

Reply:
(Optional) 
@jyoungyun 
The code `std::lock_guard<std::mutex> guard(lock);` can be moved after calling `CodegenLoader::instance()`. `CodegenLoader::instance()` is thread-safe as you know,.

Reply:
According to you(@jyoungyun and @ragmani ), I can learn so many things about `mutex`. Thank you! 👍

Reply:
> @glistening
> 
> Could you let us know which code is not thread-safe? I checked `CodegenLoader::instance()` is thread-safe at [#12553 (comment)](https://github.com/Samsung/ONE/pull/12553#discussion_r1470677581).

No. I am not saying `Codegen...` is thread-unsafe. The generator library (`libtvngen.so`) is not thread-safe.